src/easylink/pipeline_graph.py

```diff
@@ -47,53 +50,81 @@ def update_slot_filepaths(self, config: Config) -> None:
         ):
             for edge_idx in self[source][sink]:
                 if edge_attrs["output_slot"].name == "all":
-                    self[source][sink][edge_idx]["filepaths"] = [
-                        str(path) for path in config.input_data.to_dict().values()
-                    ]
+                    self[source][sink][edge_idx]["filepaths"] = tuple(
+                        [str(path) for path in config.input_data.to_dict().values()]
+                    )
                 else:
-                    self[source][sink][edge_idx]["filepaths"] = [
-                        str(config.input_data[edge_attrs["output_slot"].name])
-                    ]
+                    self[source][sink][edge_idx]["filepaths"] = (
+                        str(config.input_data[edge_attrs["output_slot"].name]),
+                    )
 
         # Update implementation nodes with yaml metadata
         for node in self.implementation_nodes:
             imp_outputs = self.nodes[node]["implementation"].outputs
             for source, sink, edge_attrs in self.out_edges(node, data=True):
                 for edge_idx in self[node][sink]:
-                    self[source][sink][edge_idx]["filepaths"] = [
+                    self[source][sink][edge_idx]["filepaths"] = (
                         str(
                             Path("intermediate")
                             / node
                             / imp_outputs[edge_attrs["output_slot"].name]
-                        )
-                    ]
+                        ),
+                    )
 
-    def get_input_slots(self, node: str) -> Dict[str, List[str]]:
+    def get_input_slots(self, node: str) -> dict[str, dict[str, Union[str, list[str]]]]:
         """Get all of a node's input slots from edges."""
-        input_slots = {}
-        for _, _, edge_attrs in self.in_edges(node, data=True):
-            # Consider whether we need duplicate variables to merge
-            env_var, files = edge_attrs["input_slot"].env_var, edge_attrs["filepaths"]
-            if env_var in input_slots:
-                input_slots[env_var].extend(files)
+        input_slots = [
+            edge_attrs["input_slot"] for _, _, edge_attrs in self.in_edges(node, data=True)
+        ]
+        filepaths_by_slot = [
+            list(edge_attrs["filepaths"])
```

Comment:
Why are you converting this tuple back to a list? It seems like passing a tuple to `condense_input_slots` should work as well.

Reply:
I'm gathering together the filepaths by using `.extend(filepaths)` later, and I thought it was more convenient to convert to list now instead of in `condense_input_slots`